fe1-web/network/NetworkConnexion.ts

```diff
@@ -0,0 +1,109 @@
+import { w3cwebsocket as W3CWebSocket } from 'websocket';
+import { JsonRpcRequest, JsonRpcResponse, UNDEFINED_ID } from 'model/network';
+
+const GENERATE_ID: number = -1;
+
+const WEBSOCKET_READYSTATE_INTERVAL_MS = 10;
+const WEBSOCKET_READYSTATE_MAX_ATTEMPTS = 100;
+
+export class NetworkConnexion {
+  private readonly ws: any;
+
+  private payloadSentCount: number = 0;
+
+  private payloadPending: Map<number, Promise<JsonRpcResponse>>;
+
+  public readonly address: string; // FIXME get the address from ws (log ws when app is running)
+
+  constructor(address: string, onMessageHandler: (message: any) => void) {
+    const ws = new W3CWebSocket(address);
+
+    ws.onopen = () => { console.info(`initiating web socket : ${address}`); };
+    ws.onmessage = (message: any) => {
+      console.log(`Received a new message from '${address}' : `, message.data);
+
+      let response: JsonRpcResponse;
+      try {
+        response = JsonRpcResponse.fromJson(message.data);
+      } catch (e) {
+        throw new Error('Answer is not valid Json');
+      }
+
+      if (response.id !== UNDEFINED_ID) {
+        const promise = this.payloadPending.get(response.id);
+        this.payloadPending.delete(response.id);
+
+        // FIXME resolve the promise
+      }
+
+      onMessageHandler(response);
+    };
+    ws.onerror = (event: any) => { console.error(`WebSocket error observed on '${address}' : `, event); };
+
+    this.address = address;
+    this.payloadPending = new Map();
+    this.ws = ws;
+  }
+
+  public sendPayload(payload: JsonRpcRequest): void {
+    // Check that the websocket connection is ready
+    if (!this.ws.readyState) {
+      new Promise((resolve, reject) => {
+        this.waitWebsocketReady(resolve, reject);
+      }).then(
+        () => this.sendPayload(payload),
+      );
+    } else {
+      let query: JsonRpcRequest = payload;
+
+      // websocket ready to be used, message can be sent
+      if (payload.id === undefined || payload.id === GENERATE_ID) {
```

Comment:
this is kind of dangerous, because the application could generate an ID that we've already used in a RPC in progress.

I'm thinking that maybe we could "force" the application to use "GENERATE_ID" and return an error if they don't, but always overwrite the `id` field here.

Reply:
This was a mechanism implemented back in the time to be able to send the same query another time (in case of server fault) with the same id.
```pseudocode
let query = // get the query from pending queue
if (retry) {
     send(id = query.id, payload = query);
} else {
    send(id = GENERATE_ID, payload = query);
}
```

However if we return promises back to the user (component), then this wouldn't be useful anymore in my opinion 🤔 . What do you think about it @pierluca ?

Reply:
@RaulinN  If there was a fault during the first query, it won't hurt the system to do the same RPC (again) with a different request id.

Reply:
Yes indeed! That's what I was thinking if we still need to send the same query! I just explained it poorly in my last message ^^. So I agree with you, no need to let the user choose the id anymore